### PR TITLE
[Merged by Bors] - feat: counting elements in an interval with given residue

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Int/Card.lean [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+Pass
+-/
+
+
+open Finset Int
+
+/-- There are `⌈b / r⌉ - ⌈a / r⌉` multiples of `r` in `[a, b)`, if `a ≤ b`. -/
+theorem Int.Ico_filter_dvd_card (a b : ℤ) {r : ℤ} (hr : 0 < r) :
+    ((Ico a b).filter (r ∣ ·)).card = max (⌈b / (r : ℚ)⌉ - ⌈a / (r : ℚ)⌉) 0 := by
```
Comment (eric-wieser) [inline Mathlib/Data/Int/Card.lean:22]: I think you should actually prove
```suggestion
    (Ico a b).filter (r ∣ ·) = (Ico ⌈b / (r : ℚ)⌉ ⌈a / (r : ℚ)⌉).map (.mul_right) := by
```
Or similar

Comment (Parcly-Taxel) [reply Mathlib/Data/Int/Card.lean:22]: That was in fact how my original proof went (_separately_ establishing equality of sets), but @b-mehta (when I presented the declaration for golfing on Xena) thought that _integrating_ the equivalence proof into the cardinality theorem through `card_congr` was better. I agreed with him.

**Comment by Parcly-Taxel** [comment]
Also, is there a better place to put these theorems?

File: Mathlib/Data/Int/Card.lean (hunk 1 above)
Comment (Parcly-Taxel) [reply Mathlib/Data/Int/Card.lean:29]: …never mind, your suggestion turns out to be even shorter.

File: Mathlib/Data/Int/Card.lean [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+# Number of elements in an integer/natural number interval congruent to a given number
+
+Pass
+-/
+
+
+open Finset Int
+
+theorem Int.Ico_filter_dvd_eq (a b : ℤ) {r : ℤ} (hr : 0 < r) : (Ico a b).filter (r ∣ ·) =
+    (Ico ⌈a / (r : ℚ)⌉ ⌈b / (r : ℚ)⌉).map ⟨_, mul_left_injective₀ hr.ne'⟩ := by
```
Comment (eric-wieser) [inline Mathlib/Data/Int/Card.lean:21]: Why not
```suggestion
    (Ico (a / r) (b / r)).map ⟨_, mul_left_injective₀ hr.ne'⟩ := by
```
Am I missing an obvious reason why you need to round up and not down?

Comment (eric-wieser) [reply Mathlib/Data/Int/Card.lean:21]: Ah, I guess the open endpoint has to round up, but the closed end doesn't

Comment (eric-wieser) [reply Mathlib/Data/Int/Card.lean:21]: Ah nevermind, I see why

Comment (eric-wieser) [reply Mathlib/Data/Int/Card.lean:21]: Can you add the Ioc version too?

Comment (Parcly-Taxel) [reply Mathlib/Data/Int/Card.lean:27]: Added!

File: Mathlib/Data/Int/Card.lean
[the remark this answers is not in the document]
Comment (Parcly-Taxel) [reply Mathlib/Data/Int/Card.lean:1]: I don't know a better name – or home – for this actually.

[the remark this answers is not in the document]
Comment (Parcly-Taxel) [reply Mathlib/Data/Int/Card.lean:1]: @YaelDillies do you have any advice on this matter?

[the remark this answers is not in the document]
Comment (sgouezel) [reply Mathlib/Data/Int/Card.lean:1]: CardIntervalMod, for instance?

File: Mathlib/Data/Int/Card.lean [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+
+open Finset Int
+
+namespace Int
+
+variable (a b : ℤ) {r : ℤ} (hr : 0 < r)
+
+lemma Ico_filter_dvd_eq : (Ico a b).filter (r ∣ ·) =
+    (Ico ⌈a / (r : ℚ)⌉ ⌈b / (r : ℚ)⌉).map ⟨_, mul_left_injective₀ hr.ne'⟩ := by
```
Comment (sgouezel) [inline Mathlib/Data/Int/Card.lean:28]:
```suggestion
    (Ico ⌈a / (r : ℚ)⌉ ⌈b / (r : ℚ)⌉).map ⟨(· * r), mul_left_injective₀ hr.ne'⟩ := by
```
It doesn't help Lean, but it definitely helps the reader understand what the theorem is saying.

Same thing below several times.

File: Mathlib/Data/Int/Card.lean [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+lemma Ioc_filter_modEq_cast {v : ℕ} : ((Ioc a b).filter (· ≡ v [MOD r])).map castEmbedding =
+    (Ioc ↑a ↑b).filter (· ≡ v [ZMOD r]) := by
+  ext x
+  simp only [mem_map, mem_filter, mem_Ioc, castEmbedding_apply]
+  constructor
+  · simp_rw [forall_exists_index, ← coe_nat_modEq_iff]; intro y ⟨h, c⟩; subst c; exact_mod_cast h
+  · intro h; lift x to ℕ using (by linarith); exact ⟨x, by simp_all [coe_nat_modEq_iff]⟩
+
+/-- `Int.Ico_filter_modEq_card` restricted to natural numbers. -/
```
Comment (sgouezel) [inline Mathlib/Data/Int/Card.lean:97]: This docstring is not really helpful, as the reader needs to hunt what the other lemma is saying (and tooltips won't help). Could you instead explain the content of the lemma in the docstring? 

Same thing below.

Comment (Parcly-Taxel) [reply Mathlib/Data/Int/Card.lean]: Done.

**Comment by sgouezel** [comment]
bors r+
Thanks!

**Pull request closed without merging**